lxd/storage/backend_lxd.go

```diff
@@ -855,6 +855,15 @@ func (b *lxdBackend) EnsureImage(fingerprint string, op *operations.Operation) e
 		return nil // Nothing to do for drivers that don't support optimized images volumes.
 	}
 
+	// We need to lock this operation to ensure that the image is not being
+	// created multiple times.
+	imageLockID := fmt.Sprintf("create/%s/%s/%s", b.name, drivers.VolumeTypeImage, fingerprint)
```

Comment:
@monstermunchkin I was thinking about this locking mechanism, that the way the lock label is generated might lead to issues in the future around understanding the lock label format and/or repetition leading to typos in generating it.

How about rather than a `Lock(ID)` function that the signature to be changed to something like `Lock(action, poolName, volType, volName)` that way the internal encoding mechanism to derive the key can be encapsulated and it is clearer to the user of the Lock function what it is that needs to be supplied.

Also, with regards to the action part of the lock, I wonder if this is truly needed, or whether perhaps a 'level' argument might be better suited, or no argument at all.

For example there might be the scenario where 2 different actions are occurring concurrently on the same volume that would interfere with each other (for example mounting and renaming, or mounting and unmounting) that should really lock each other, however as they would likely be using a different action argument they would not. Perhaps it would be simpler for the time being to just use the lock to establish exclusive access to a volume to do what is needed, and block all other operations from occurring on it.

CC @stgraber in case he has any insight into this.

Reply:
Yeah, locking the entire volume seems less likely to cause issues. If we notice some kind of performance issue as a result of that, we can try to go more fine grained.